### PR TITLE
TypeScript fixes again

### PR DESCRIPTION
**Pull request by thw0rted**
This is an updated branch for #8903.  I hate to lose the conversation from there but submitting these changes as a single commit allowed me to significantly reduce the number of files in the patchset -- a lot of them were only related to `when.js` Promises, and that's no longer relevant.  Hopefully this makes it easier to review and merge.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @thw0rted!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/more-typescript-redo/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.

**Comment by thw0rted** [comment]
I just noticed that #10265 was called out in the changelog -- turns out that two of the "fixes" I made to the return type info should have been correct and it was the code that was wrong.  It might be worth looking at the function bodies for the other return-type changes I have in this PR to double check that there aren't other similar errors waiting to bite somebody.

File: Source/Core/HeightmapTerrainData.js [hunk 1]
```diff
@@ -500,9 +500,8 @@ HeightmapTerrainData.prototype.interpolateHeight = function (
  * @param {Number} descendantX The X coordinate within the tiling scheme of the descendant tile for which we are upsampling.
  * @param {Number} descendantY The Y coordinate within the tiling scheme of the descendant tile for which we are upsampling.
  * @param {Number} descendantLevel The level within the tiling scheme of the descendant tile for which we are upsampling.
- * @returns {Promise.<HeightmapTerrainData>|undefined} A promise for upsampled heightmap terrain data for the descendant tile,
- *          or undefined if too many asynchronous upsample operations are in progress and the request has been
- *          deferred.
+ * @returns {HeightmapTerrainData|undefined} The upsampled heightmap terrain data for the descendant tile,
```
Comment (thw0rted) [inline Source/Core/HeightmapTerrainData.js:503]: Likely another Promise bug like #10265

File: Source/Core/buildModuleUrl.js [hunk 2]
```diff
@@ -5,6 +5,7 @@ import Resource from "./Resource.js";
 
 /*global CESIUM_BASE_URL,define,require*/
 
+// TODO: all these should have JSDoc tags, or mark as private
```
Comment (thw0rted) [inline Source/Core/buildModuleUrl.js:8]: I think this TODO was intended to be attached to the two property-functions (?) attached to the exported `buildModuleUrl`.  Basically, you can `import buildModuleUrl from "./buildModuleUrl.js"`, then call `buildModuleUrl.setBaseUrl(...)` or `buildModuleUrl.getCesiumBaseUrl()`, but the JSDoc for two functions is not recognized and doesn't show up in either the generated "globals" documentation, nor in the TS definitions.

My personal recommendation would to refactor the module to stop using `export default` and produce three named exports instead, but that's a significant breaking change that the team would have to discuss.

File: Source/Scene/ImageryProvider.js [hunk 3]
```diff
@@ -296,7 +296,7 @@ ImageryProvider.prototype.getTileCredits = function (x, y, level) {
  * @param {Number} y The tile Y coordinate.
  * @param {Number} level The tile level.
  * @param {Request} [request] The request object. Intended for internal use only.
- * @returns {Promise.<HTMLImageElement|HTMLCanvasElement>|undefined} A promise for the image that will resolve when the image is available, or
+ * @returns {HTMLImageElement|HTMLCanvasElement|ImageBitmap|Promise.<HTMLImageElement|HTMLCanvasElement|ImageBitmap>|undefined} A promise for the image that will resolve when the image is available, or
```
Comment (thw0rted) [inline Source/Scene/ImageryProvider.js:299]: Per #10265 , is this an undesired change?  Or would it be better if the caller had to expect to handle both immediate value-returns as well as a returned Promise?  (If y'all switched over to async/await this would be a lot simpler...)

Comment (thw0rted) [reply Source/Scene/ImageryProvider.js:299]: Also the issue would have been much more visible if you used actual ES6 classes with inheritance rather than having "ImageryProviders" that don't actually inherit from ImageryProvider, not sure if that's still on the table....

File: Source/Scene/ImageryProvider.js [hunk 4]
```diff
@@ -338,6 +338,10 @@ ImageryProvider.prototype.pickFeatures = function (
 
 const ktx2Regex = /\.ktx2$/i;
 
+// TODO: This can definitely return a (Promise for) ImageBitmap, and also
```
Comment (thw0rted) [inline Source/Scene/ImageryProvider.js:341]: I guess this is really a continuation of my question above.  Basically the documented return type is at least incomplete, and might actually be wrong as well.  If all ImageryProviders inherited from ImageryProvider, this would be caught automatically.

File: Source/Scene/SingleTileImageryProvider.js [hunk 5]
```diff
@@ -454,10 +454,7 @@ SingleTileImageryProvider.prototype.getTileCredits = function (x, y, level) {
  * @param {Number} y The tile Y coordinate.
  * @param {Number} level The tile level.
  * @param {Request} [request] The request object. Intended for internal use only.
- * @returns {Promise.<HTMLImageElement|HTMLCanvasElement>|undefined} A promise for the image that will resolve when the image is available, or
- *          undefined if there are too many active requests to the server, and the request
- *          should be retried later.  The resolved image may be either an
- *          Image or a Canvas DOM object.
+ * @returns {HTMLImageElement|ImageBitmap} The resolved image
```
Comment (thw0rted) [inline Source/Scene/SingleTileImageryProvider.js:457]: Probably the error from #10265 again

File: Source/Core/buildModuleUrl.js (hunk 2 above)
Comment (ggetz) [reply Source/Core/buildModuleUrl.js:8]: I'd appreciate if you can open a new issue for this. There's already some discussion around whether this function is intended for public use.

File: Source/DataSources/CallbackProperty.js [hunk 6]
```diff
@@ -104,8 +104,8 @@ CallbackProperty.prototype.equals = function (other) {
  * A function that returns the value of the property.
  * @callback CallbackProperty.Callback
  *
- * @param {JulianDate} [time] The time for which to retrieve the value.
- * @param {Object} [result] The object to store the value into, if omitted, a new instance is created and returned.
- * @returns {Object} The modified result parameter or a new instance if the result parameter was not supplied or is unsupported.
+ * @param {JulianDate} time The time for which to retrieve the value.
```
Comment (ggetz) [inline Source/DataSources/CallbackProperty.js:107]: I think the `time` parameter _is_ optional.

File: Source/DataSources/CompositePositionProperty.js [hunk 7]
```diff
@@ -87,7 +87,7 @@ Object.defineProperties(CompositePositionProperty.prototype, {
  *
  * @param {JulianDate} time The time for which to retrieve the value.
  * @param {Object} [result] The object to store the value into, if omitted, a new instance is created and returned.
```
Comment (ggetz) [inline Source/DataSources/CompositePositionProperty.js:89]: This should also be `Cartesian3`, right?

File: Source/Core/ScreenSpaceEventHandler.js [hunk 8]
```diff
@@ -934,12 +934,13 @@ function ScreenSpaceEventHandler(element) {
   registerListeners(this);
 }
 
+// TODO: document callback signature for each type of event
```
Comment (ggetz) [inline Source/Core/ScreenSpaceEventHandler.js:937]: Do you plan on tackling this TODO?

File: Source/DataSources/TimeIntervalCollectionPositionProperty.js [hunk 9]
```diff
@@ -57,6 +57,7 @@ Object.defineProperties(TimeIntervalCollectionPositionProperty.prototype, {
    * Gets the interval collection.
    * @memberof TimeIntervalCollectionPositionProperty.prototype
    * @type {TimeIntervalCollection}
+   * @readonly
```
Comment (ggetz) [inline Source/DataSources/TimeIntervalCollectionPositionProperty.js:60]: `@readonly` applies to `intervals` as well, correct?

File: Source/Scene/ImageryProvider.js (hunk 3 above)
Comment (ggetz) [reply Source/Scene/ImageryProvider.js:299]: I think we want to have each `ImageryProvider` implementation return a `Promise` or `undefined`. Yes, I agree it's difficult to enforce with the current state of things, but full migration to ES6 classes is a bit out of scope for this PR. 😅

**Review by ggetz — COMMENTED** [review]
Thanks @thw0rted! I have a few comments. It appears that most of the typescript-specific question have been discussed back in https://github.com/CesiumGS/cesium/pull/8903, so mine comments are mostly about the documentation itself.

File: Source/Core/buildModuleUrl.js (hunk 2 above)
Comment (thw0rted) [reply Source/Core/buildModuleUrl.js:8]: Hrm.  I found #8922 and #9783 so it looks like your team was already looking at / working on this.  Maybe you should re-open one or both of those until the discussion is finished?

File: Source/DataSources/CallbackProperty.js (hunk 6 above)
Comment (thw0rted) [reply Source/DataSources/CallbackProperty.js:107]: This is a kind of a mess.

Since this is a callback-function signature, what we're actually saying here is whether or not the function is guaranteed to be invoked with a time value, or whether the function must handle the case where `time` is undefined.  The callback is invoked by passing through the `time` argument when `getValue` is called.

[Here](https://github.com/thw0rted/cesium/blob/more-typescript-redo/Source/DataSources/CallbackProperty.js#L52) is the `time` argument for `CallbackProperty#getValue` -- it looks like this was erroneously copy-pasted from `ConstantProperty`, since obviously a ConstantProperty does not vary with time but a CallbackProperty (probably?) could.  *If* the `time` argument for CallbackProperty should in fact be required, then the callback signature can be marked as non-optional, and callbacks can safely assume `time` is not undefined.  My inclination is to resolve this by marking the `time` argument to `getValue` as required as well.  That would be in line with the [argument in the base Property interface](https://github.com/thw0rted/cesium/blob/more-typescript-redo/Source/DataSources/Property.js#L55), which is required.

File: Source/Scene/ImageryProvider.js (hunk 4 above)
Comment (thw0rted) [reply Source/Scene/ImageryProvider.js:341]: I'm trying to address this by exporting a typedef to represent all the types an ImageryProvider might return.  So, providers that punt to `ImageryProvider.loadImage` can just resolve as the typedef, but providers that always return one known type can be more specific in their return annotation.

File: Source/DataSources/CompositePositionProperty.js (hunk 7 above)
Comment (thw0rted) [reply Source/DataSources/CompositePositionProperty.js:89]: Good catch!

File: Source/DataSources/TimeIntervalCollectionPositionProperty.js (hunk 9 above)
Comment (thw0rted) [reply Source/DataSources/TimeIntervalCollectionPositionProperty.js:60]: You might have been looking at the next field, `referenceFrame` -- yes, that should also be `@readonly`, should be fixed in the changes I'm about to push.

File: Source/Core/ScreenSpaceEventHandler.js (hunk 8 above)
Comment (thw0rted) [reply Source/Core/ScreenSpaceEventHandler.js:937]: It's a little complicated.  I'm making typedefs and a callback to describe the event types that can be passed to a callback, but the right way to solve it involves a number of overloads for `setInputAction`, at least one per type of callback (of which there are five).  JSDoc overloads don't work very well (see previous PR with `exportKML`), and cause issues with `prettier`.

For now, I'm going to push out a version with a type-union callback parameter.  I'll look at doing overloads but will probably hold off unless/until there's a better way to implement them.

Comment (thw0rted) [reply Source/Core/ScreenSpaceEventHandler.js]: OK I got this 100% working, I'm pretty sure, but it is going to take a little explaining.  Overloads in JSDoc are a bit hairy; there are pros and cons to my approach.

On the pro side, Typescript users can now write `viewer.screenSpaceEventHandler.setInputAction(evt => doStuff(Cartographic.fromCartesian(evt.position)), ScreenSpaceEventType.MOUSE_DOWN)` and the `position` property will be correctly typed.  This is pretty great!  It's the same way that TS lets you write e.g. `input.addEventListener("input", evt => doStuff(evt.inputType))` and get strong typing where `evt` is automatically an `InputEvent`.

On the con side, there's a bunch of extra typedefs and `@callback`s that shouldn't really need to be named.  It takes up extra space in the source file (though it's all comments so it should minify out for production) and adds some entries to the docs of questionable utility.

I did update the "container" template in the docs so that any function declaration with a `@variation` tag greater than 1 is not printed in the docs.  This avoids a problem described in the previous version of this PR where the overridden function would get one entire doc entry for each override -- `@variation 1` is the one that gets printed.  (I just found `Documentation\Contributors\DocumentationGuide`, will add a quick description of this behavior there!)

File: Source/Core/buildModuleUrl.js (hunk 2 above)
Comment (ggetz) [reply Source/Core/buildModuleUrl.js:8]: Re-opened https://github.com/CesiumGS/cesium/issues/8922.

File: Source/DataSources/CallbackProperty.js (hunk 6 above)
Comment (ggetz) [reply Source/DataSources/CallbackProperty.js:107]: OK, I think I understand your point, thanks! I agree the parameter for the `getValue` function should be required.

File: Source/Core/ScreenSpaceEventHandler.js (hunk 8 above)
Comment (ggetz) [reply Source/Core/ScreenSpaceEventHandler.js]: I have to say I'm a bit skeptical about this approach. @kring, would you be able to weigh in here?

Comment (thw0rted) [reply Source/Core/ScreenSpaceEventHandler.js]: If we need to back this out, let me know.  I really dislike the syntax we're forced to use to implement overloading via (tsd-)JSDoc; in TypeScript it would be one line per overload instead of 10+, and it's *much* easier to read.

That said, the resulting overloaded types really are quite helpful, especially when registering event handlers via a one-liner lambda function.  The alternative is that callbacks always type as a plain `Function` (`(...args: any[]) => void`) which, when the TS compiler is in "strict" mode, means that every callback has to explicitly type the arguments.  Worse, as far as I can tell, they aren't actually written out explicitly anywhere in (the current version of) the docs.

Comment (ggetz) [reply Source/Core/ScreenSpaceEventHandler.js]: Thanks @thw0rted, let's back it out so we can focus on getting the other improvements in this PR merged.

Feel free to open an issue or a separate PR for the overloaded types, and we can address them separately. :+1:

Comment (thw0rted) [reply Source/Core/ScreenSpaceEventHandler.js]: Please take a look and let me know if it's what you had in mind.  (I also caught one more little error, I just keep this branch open and correct stuff when it bites me.)

File: Source/Core/buildModuleUrl.js (hunk 2 above)
Comment (ggetz) [inline Source/Core/buildModuleUrl.js:8]: To get this merged, there shouldn't be any `TODO` comments in the code. Any action item should be in an issue so we can track them. Or let me know if there was discussion around any of these that wasn't resolved.

**Comment by ggetz** [comment]
Thanks @thw0rted! After the last commit, the `build-ts` task is failing in CI. Could you please take a look?

File: Source/Core/buildModuleUrl.js (hunk 2 above)
Comment (thw0rted) [reply Source/Core/buildModuleUrl.js]: I took out the comment, should be addressed by https://github.com/CesiumGS/cesium/issues/8922 which you already re-opened.

File: Source/Scene/ImageryProvider.js [hunk 10]
```diff
@@ -338,17 +347,19 @@ ImageryProvider.prototype.pickFeatures = function (
...
```
[the remark this answers is not in the document]
Comment (thw0rted) [reply Source/Scene/ImageryProvider.js:350]: No sweat, I missed that too.  I was about to remove it, but should ask first, do you want the return type for the method (`loadImage`) to include `CompressedTextureBuffer`?  It looks like that's what you get if you pass it a `ktx2` file.

[the remark this answers is not in the document]
Comment (ggetz) [reply Source/Scene/ImageryProvider.js:350]: Yes, please add `CompressedTextureBuffer`. Thanks!

[the remark this answers is not in the document]
Comment (thw0rted) [reply Source/Scene/ImageryProvider.js:350]: Fixed.  I also resolved a couple very minor warnings from `tsd-jsdoc` that I noticed while testing this change, mostly self explanatory.  The `@private` ModelExperimental property is tagged private in a sibling class so I assumed it was an oversight.  The change from `String` to `string` is an ongoing effort to stop claiming to require/return boxed primitives when that's not actually true. (It's never actually true...)

**Comment by ggetz** [comment]
Thanks @thw0rted! Latest changes look good. Can you back out the `Documentation/` changes since they're no longer needed? Then this is ready to be merged. :+1:

File: Source/Core/buildModuleUrl.js [hunk 11]
```diff
@@ -142,6 +142,9 @@ buildModuleUrl.setBaseUrl = function (value) {
 
 /**
  * Gets the base URL for resolving modules.
+ *
+ * @function
+ * @returns {string} the configured base URL
```
Comment (ggetz) [inline Source/Core/buildModuleUrl.js:147]: We do use capital `S` `String` throughout the codebase. Is there a reason that's changed in this file?

Comment (thw0rted) [reply Source/Core/buildModuleUrl.js:147]: This is one of those questions where I convinced myself of the answer years ago but forgot most of the justification in the meantime.  [Typescript is very opinionated about it](https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html), but doesn't really offer much justification in their docs.  There's some good discussion on [this SO question](https://stackoverflow.com/questions/6793228/does-a-matter-whether-a-value-is-primitive-or-boxed).  (If you poke around on SO, you'll also find questions about "why do boxed types even exist?" with quotes from the language designers blaming management and generally regretting the decision.)

The short version is: lowercase-s `string` is a primitive and title-case `String` is an object ("box").  The most visible difference called out in that SO answer is that `typeof new String()` is `object`, not `string`. Primitives are boxed automatically if needed, but if only a primitive will do, you don't get free un-boxing.  I think the idea is that you probably almost never return `new String` in your code, but you might do a `typeof` check on a function argument at runtime, so the safest thing is to always use the primitive type in docs.

**Comment by thw0rted** [comment]
Re the `Documentation` changes (and the corresponding JSDoc template update), does this mean that the team isn't willing to allow for any function overloading?  I think I had this conversation with Matt on the previous PR, several years ago, but at the time I hadn't figured out a way to avoid having multiple entries in the actual docs -- which is of course pretty confusing.

5+ overloads in one file looks bad, but the previous case was just one, I think so that exportKML gave the correct return type based on the options with which it was called.  That would be a lot less verbose/confusing with the new system.  I guess in the worst case I can always dig up the old commit if we need it again, but it'd be nice to leave it in place as an option.  Your call?

**Comment by ggetz** [comment]
I would say hold off on adding the guide and the tooling until we have a verified use case in the code. It may be the path we end up going down eventually, and we'll have the record of it in this PR. But until we have a use case in the code itself, maintaining the documentation and tolling is questionable.

File: Source/Core/buildModuleUrl.js (hunk 11 above)
Comment (thw0rted) [reply Source/Core/buildModuleUrl.js:147]: I guess we could back it out?  I don't see the harm in making piecemeal changes, at least if we're touching adjacent docs (as I was in this case).  Not a big deal one way or another.

Comment (ggetz) [reply Source/Core/buildModuleUrl.js:147]: Yes, thanks. I think we prefer the consistency. This is also how we've approach code style changes to the code base: All at once, ideally automating the change where possible.

**Comment by ggetz** [comment]
Awesome, thanks @thw0rted! Thanks for your patience throughout the process.
